v2/traversal/resumption.go

```diff
@@ -17,12 +17,41 @@ import (
 	"github.com/ipld/go-ipld-prime/traversal"
 )
 
+// TraverseResumer allows resuming a progress from a previously encountered path in the selector.
+type TraverseResumer interface {
+	RewindToPath(from datamodel.Path) error
+	RewindToOffset(offset uint64) error
+	Position() uint64
+}
+
+// TraversalResumerPathState tracks a traversal state for the purpose of
+// building a CAR. For each block in the CAR it tracks the path to that block,
+// the Link of the block and where in the CAR the block is located.
+//
+// A TraversalResumerPathState shared across multiple traversals using the same
+// selector and DAG will yield the same state. This allows us to resume at
+// arbitrary points within in the DAG and load the minimal additional blocks
+// required to resume the traversal at that point.
+type TraversalResumerPathState interface {
+	AddPath(path []datamodel.PathSegment, link datamodel.Link, atOffset uint64)
+	GetLinks(root datamodel.Path) []datamodel.Link
+	GetOffsetAfter(root datamodel.Path) (uint64, error)
+}
+
 type pathNode struct {
 	link     datamodel.Link
 	offset   uint64
 	children map[datamodel.PathSegment]*pathNode
 }
 
+// NewTraversalResumerPathState creates a new TraversalResumerPathState.
+//
+// Note that the TraversalResumerPathState returned by this factory is not
+// thread-safe.
+func NewTraversalResumerPathState() TraversalResumerPathState {
+	return newPath(nil, 0)
+}
+
```

Comment:
do we still need `newPath` below or can we just have this?

Reply:
it's called again in `AddPath` with properties, so it's easier to leave it like this I think